lnwire/custom_records.go

```diff
@@ -0,0 +1,52 @@
+package lnwire
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+const (
+	// MinCustomRecordsTlvType is the minimum custom records TLV type as
+	// defined in BOLT 01.
+	MinCustomRecordsTlvType = 65536
+)
+
+// CustomRecords stores a set of custom key/value pairs.
+type CustomRecords map[uint64][]byte
+
+// NewCustomRecordsFromTlvTypeMap creates a new CustomRecords instance from a
+// tlv.TypeMap.
+func NewCustomRecordsFromTlvTypeMap(tlvMap tlv.TypeMap) (*CustomRecords,
```

Comment:
Why do we need a completely new type? Can't we just create a `ValidateCustomTypes(tlvMap tlv.TypeMap)` function that does this? Especially since this is more or less an exact copy of the `record.CustomSet` type as well.

Reply:
> Why do we frequently choose a more generic type (tlv.TypeMap) that lacks the necessary restrictiveness

I think it's generally not the best idea to have too much business/protocol logic in the wire/encoding level, as that reduces the re-usability. I know that shifts the responsibility of validating the content to the caller. But should a wire message really be dictating what its _content_ should be, rather than focusing on how to encode the content?

To me the encoding function as it looks now mostly distracts from what we actually do: merging two sets of TLV records into a single encoded blob.

Reply:
I've added a method called `CustomRecords.ExtendRecordProducers` which moves the protocol logic to the `CustomRecords` type away from the `UpdateAddHtlc.Encode` method.